app/controllers/bookmarks_controller.rb

```diff
@@ -0,0 +1,63 @@
+class BookmarksController < ApplicationController
+  def show
+  end
+
+  def new
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = @topic.bookmarks.new
+  end
+  
+  def create
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = @topic.bookmarks.build(bookmark_params)
+    
+    if @bookmark.save
+      flash[:notice] = "Bookmark created successfully"
+      redirect_to(@topic)
+    else
+      flash[:alert] = "Issue creating bookmark"
+      render :new
+    end
+  end
+
+  def edit
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = Bookmark.find(params[:id])
+  end
+  
+  def update
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = Bookmark.find(params[:id])
+    @bookmark.assign_attributes(bookmark_params)
+    
+    if @bookmark.save
+      flash[:notice] = "Bookmark saved successfully"
+      redirect_to(@topic)
+    else
+      flash[:notice] = "Error updating bookmark"
+      render :edit
+    end
+  end
+  
+  def destroy
+    @user = current_user
+    @topic = @user.topics.find(params[:topic_id])
+    @bookmark = Bookmark.find(params[:id])
+    
+    if @bookmark.destroy
+      redirect_to(@topic)
+    else
+      flash[:alert] = "There was an error deleting the bookmark"
+      redirect_to(@topic)
+    end
+  end
```

Comment:
@ianroberts131 `destroy` action also has a similar vulnerability as to those mentioned above with `update` and `edit` actions.